### PR TITLE
feat: Add mandrill base classes

### PR DESCRIPTION
**Pull request by dacgray**
### Mandrill API spec:

- https://mandrillapp.com/api/docs/messages.JSON.html#method=send-template
- https://mandrillapp.com/api/docs/messages.JSON.html#method=send

## PR

This PR provides the framework for including Mandrill emails in this package.

It adds 4 classes:

### BaseEmail

The basic email class, containing email building methods.

### RegularEmail

Inherit this class to add a regular email - ie a text or html email.

### TemplateEmail

Inherit this class to add a template email - ie an email where the template resides on Mandrill.

### SendResult

The result data returned from Mandrill after a send email attempt.

### Usage

We will add email classes in the CrazyFactory\ShopEmails\Mandrill\Email namespace.  Email classes will inherit from TemplateEmail or RegularEmail:

```php
class ForgotPasswordEmail extends TemplateEmail
{

    const TEMPLATE_NAME = 'FORGOT_PASSWORD';
    
    const MERGE_VAR_NAME = 'blah';
    
    public function __construct(
        string $apiKey,
        string $recipientEmail,
        string $recipientName,
        string $someContent,
        ?Mandrill $mandrillClient = null
    ) {
        parent::__construct($apiKey, $mandrillClient);
        
        $this->addRecipient($recipientEmail, $recipientName)
            ->addMergeVar(self::MERGE_VAR_NAME, $someContent);
    }

}
```

### Implementation

```php
$results = (new ForgotPasswordEmail(
    'apiKey',
    'test@example.com',
    'name',
    'your new password is...'
))->send();

if (current($results)->getStatus() === SendResult::STATUS_SENT) {
    // do something
}
```

### CODE REVIEW

File: src/Mandrill/Core/BaseEmail.php [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    }
+
+    /**
+     * @param string[] $filePaths
+     * @param null|string $mimeType
+     *
+     * @return $this
+     * @throws \Exception
+     */
+    public function addAttachments(array $filePaths, ?string $mimeType = null): BaseEmail
```
Comment (turboatcrazyfactory) [inline src/Mandrill/Core/BaseEmail.php]: If I have 2 files with different type, I need to call this method 2 times.

Why don't we accept file as array of array then ?
like
```php
[
     ['path' => 'filepath1', 'alias' => '1', 'mimetype' => 'image/png'],
     ['path' => 'filepath2', 'alias' => '2', 'mimetype' => 'text/plain'],
]
```

File: src/Mandrill/Core/BaseEmail.php [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        string $filePath,
+        string $alias = null,
+        ?string $mimeType = null
+    ): BaseEmail {
+        if (!file_exists($filePath)) {
+            throw new InvalidArgumentException("File at: `{$filePath}` does not exist");
+        }
+
+        // Check file is less than 25MB
+        if (filesize($filePath) > 26214400) {
```
Comment (turboatcrazyfactory) [inline src/Mandrill/Core/BaseEmail.php]: I think Mandrill is validating the whole message size not only attachment.
Do Mandrill client already handle this for us then ?
https://mandrill.zendesk.com/hc/en-us/articles/205582407-Does-Mandrill-Support-Attachments-

File: src/Mandrill/Core/SendResult.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    protected $eventId;
+
+    /**
+     * SendResult constructor.
+     * @param array $result
+     */
+    public function __construct(array $result)
+    {
+        $this->email = $result['email'];
```
Comment (turboatcrazyfactory) [inline src/Mandrill/Core/SendResult.php]: Should we also handle undefined index notice for other keys too ? (Like you did for `reject_reason`)

File: tests/Integration/Mandrill/RegularEmailTest.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+class RegularEmailTest extends UnitTest
+{
+
+    /**
+     * @throws \Throwable
+     */
+    public function testSend()
+    {
+        $email = new RegularEmail(
+            $GLOBALS['MANDRILL_API_KEY']
```
Comment (turboatcrazyfactory) [inline tests/Integration/Mandrill/RegularEmailTest.php]: Where did you set this key ?

File: tests/Integration/Mandrill/TemplateEmailTest.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            $this->assertTrue(
+                in_array(
+                    $result->status,
+                    [
+                        SendResult::STATUS_QUEUED,
+                        SendResult::STATUS_SENT
+                    ]
+                )
+            );
+        }
```
Comment (turboatcrazyfactory) [inline tests/Integration/Mandrill/TemplateEmailTest.php]: Should we also test failure case ? When send status is something that failed.

File: tests/Integration/Mandrill/RegularEmailTest.php (hunk 4 above)
Comment (dacgray) [reply tests/Integration/Mandrill/RegularEmailTest.php]: In the package here: `tests/_support/config.local.php`

We'll inject it from shop/erp/wherever local config on implementation.

File: tests/Integration/Mandrill/TemplateEmailTest.php (hunk 5 above)
Comment (dacgray) [reply tests/Integration/Mandrill/TemplateEmailTest.php]: I don't think it's necessary.   We have no influence over email failure return/exceptions in our code.  It's all the API wrapper or on Mandrill side - we build a SendResult array in the same way regardless.

File: src/Mandrill/Core/BaseEmail.php (hunk 2 above)
Comment (dacgray) [reply src/Mandrill/Core/BaseEmail.php]: ah, nice point.

File: src/Mandrill/Core/BaseEmail.php (hunk 1 above)
Comment (dacgray) [reply src/Mandrill/Core/BaseEmail.php]: Makes sense:

```php
/**
     * @param array $attachmentList
     * [
     *      [
     *          'filePath',
     *          'alias',
     *          'mimeType',
     *      ]
     * ]
     * @return $this
     * @throws \Exception
     */
    public function addAttachments(array $attachmentList): BaseEmail
    {
        foreach ($attachmentList as $attachment) {
            $this->addAttachment(...$attachment);
        }

        return $this;
    }
```

File: src/Mandrill/Core/SendResult.php (hunk 3 above)
Comment (dacgray) [reply src/Mandrill/Core/SendResult.php]: Not necessary ^^

It's only `reject_reason` that can be undefined:

image

**Pull request merged**